---
title: "Transcriptomics Exam: Bulk Analysis"
author: "Gabriele Marchi and Carlo Manenti"
date: '2022-06-24'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# **Bulk analysis**

### Loading the **library** needed for the analysis
```{r, results="hide"}
library(recount3)
library(edgeR)
library(ggplot2)
```


### Loading the **data**
```{r, results="hide"}
rse_brain <- readRDS('rse_brain.RDS')
rse_liver <- readRDS('rse_liver.RDS')
rse_colon <- readRDS('rse_colon.RDS')
```

We will perform the bulk analysis in 3 different tissue to extract differentially expressed genes. 
The tissue used are: Brain, Liver and Colon

First of all we will perform the analysis using the raw count table, so without filtering out  rRNA, mRNA, pseudogenes and non canonical chromosomes. 
Then we will perform a clean analysis filter out the aformentioned source of variability. 

The idea is to see if the methods applied to perform the analysis are robust enough to be reliable even with additional source of variations. 
Also we want to investigate if the methods are able to find meaningful differentially expressed between the three conditions. 


## **Full Bulk Analysis**

First of all we need to get the counts for each gene form the coverage
```{r}
assays(rse_brain)$counts <- transform_counts(rse_brain)
assays(rse_liver)$counts <- transform_counts(rse_liver)
assays(rse_colon)$counts <- transform_counts(rse_colon)
```


Then we can **pick the replicates**. We chose repicates according to RIN (> 6), fraction of reads of rRNA ( < 0.1) and % uniquely mapped reads ( > 85%). 

RIN >= 6 (it would be preferable to have a RIN higher than 7)
fraction of rRNA reads < 0.1 
% uniquely mapped reads > 85%

Brain:
```{r}
for (id in c(90, 91, 92)){
  cat('Sample n°', id, '\n')
  cat('RIN:', colData(rse_brain)$gtex.smrin[id], '\n')
  cat('rRNA:', colData(rse_brain)$gtex.smrrnart[id], '\n')
  cat('Mapping:', colData(rse_brain)$"recount_qc.star.uniquely_mapped_reads_%_both"[id], '\n\n')}
```
Liver:
```{r}
for (id in c(91, 94, 95)){
  cat('Sample n°', id, '\n')
  cat('RIN:', colData(rse_liver)$gtex.smrin[id], '\n')
  cat('rRNA:', colData(rse_liver)$gtex.smrrnart[id], '\n')
  cat('Mapping:', colData(rse_liver)$"recount_qc.star.uniquely_mapped_reads_%_both"[id], '\n\n')}
```
We exluded the replicates 90, 92, 93 due to a low RIN 

Colon:
```{r}
for (id in c(90, 91, 92)){
  cat('Sample n°', id, '\n')
  cat('RIN:', colData(rse_colon)$gtex.smrin[id], '\n')
  cat('rRNA:', colData(rse_colon)$gtex.smrrnart[id], '\n')
  cat('Mapping:', colData(rse_colon)$"recount_qc.star.uniquely_mapped_reads_%_both"[id], '\n\n')}
```


## **Selecting the replicates for each tissue**
```{r}
rse_brain_selected <- rse_brain[, c(90, 91, 92)]
rse_liver_selected <- rse_liver[, c(91, 94, 95)]
rse_colon_selected <- rse_colon[, c(90, 91, 92)]
```


Now we can extract the counts for each of the replicate
```{r}
counts_brain_selected <- assays(rse_brain_selected)$counts
counts_liver_selected <- assays(rse_liver_selected)$counts
counts_colon_selected <- assays(rse_colon_selected)$counts
```

## **Count Table**
Furthermore, we can build the count table and assign the official gene name.
```{r}
count.table <- cbind(counts_brain_selected, counts_liver_selected, counts_colon_selected) 

colnames(count.table) <- c('Brain.90', 'Brain.91', 'Brain.92', 'Liver.91', 'Liver.94', 'Liver.95', 'Colon.90', 'Colon.91', 'Colon.92')

rownames(count.table) <- rowData(rse_brain_selected)$gene_name 

y <- DGEList(counts =  count.table)
```

We can also define how replicates are grouped and add this information to the result object (y): 
```{r}
group <- as.factor(c('Brain', 'Brain', 'Brain', 'Liver', 'Liver', 'Liver', 'Colon', 'Colon', 'Colon'))

y$samples$group <- group
```

'Quality' information can be added to the result object for each sample: 
```{r}
# RIN 
y$samples$rin <- as.factor(c(
  colData(rse_brain_selected)$gtex.smrin, 
  colData(rse_liver_selected)$gtex.smrin, 
  colData(rse_colon_selected)$gtex.smrin))

# rRNA
y$samples$rRNA <- as.factor(c(
  colData(rse_brain_selected)$gtex.smrrnart, 
  colData(rse_liver_selected)$gtex.smrrnart, 
  colData(rse_colon_selected)$gtex.smrrnart))

# UNIQUELY MAPPED READS
y$samples$mapped <- as.factor(c(
  colData(rse_brain_selected)$'recount_qc.star.uniquely_mapped_reads_%_both',
  colData(rse_liver_selected)$'recount_qc.star.uniquely_mapped_reads_%_both', 
  colData(rse_colon_selected)$'recount_qc.star.uniquely_mapped_reads_%_both'))

# CHR % of reads mapping on the mitochondrial chromosome
y$samples$chrM <- as.factor(c(
  colData(rse_brain_selected)$'recount_qc.aligned_reads%.chrm',
  colData(rse_liver_selected)$'recount_qc.aligned_reads%.chrm', 
  colData(rse_colon_selected)$'recount_qc.aligned_reads%.chrm'))

# SLICE 
y$samples$slice <- as.factor(c(
  colData(rse_brain_selected)$gtex.smtsd, 
  colData(rse_liver_selected)$gtex.smtsd, 
  colData(rse_colon_selected)$gtex.smtsd))

# SEX 
y$samples$sex <- as.factor(c(
  colData(rse_brain_selected)$gtex.sex, 
  colData(rse_liver_selected)$gtex.sex, 
  colData(rse_colon_selected)$gtex.sex))

# AGE
y$samples$age <- as.factor(c(
  colData(rse_brain_selected)$gtex.age, 
  colData(rse_liver_selected)$gtex.age, 
  colData(rse_colon_selected)$gtex.age))
```


For a quick check we can look at how many genes have 0 counts
```{r}
table(rowSums(y$counts == 0) == 9)
```
## **Filtering by expression**
And than we can filter by expression
```{r}
keep.expressed <- filterByExpr(y, group = group)
y <- y [keep.expressed, , keep.lib.sizes = F]
```

## **logCPM**
### Transforming counts in Counts Per Milion in a Logarithmich scale (logCPM)
```{r}
logcpm_before <- cpm(y, log = T)
y <- calcNormFactors(y, method = 'TMM')
head(y)

logcpm <- cpm(y, log = T)
logcpm_after <- logcpm
```

We can visualize the normalization applied to the logCPM using box-plots

Graph of logCPM before normalization
```{r, results = 'hide'}
library(tidyr)
library(dplyr)
```

```{r}
brain <- c('Brain.90', 'Brain.91', 'Brain.92')
liver <- c('Liver.91', 'Liver.94', 'Liver.95')
colon <- c('Colon.90', 'Colon.91', 'Colon.92')

logcpm_before_long <- data.frame(logcpm_before)
#pivot the data frame into a long format
logcpm_before_long$GeneID <- rownames(logcpm_before_long)
names <- colnames(logcpm_before_long)[1: length(colnames(logcpm_before_long)) - 1]

logcpm_before_long  <- logcpm_before_long  %>% pivot_longer(cols= names ,
                    names_to= 'Samples',
                    values_to='counts')
logcpm_before_long$Legend <- ifelse(logcpm_before_long$Samples %in% brain, 'Brain', ifelse(logcpm_before_long$Samples %in% liver, 'Liver', 'Colon'))
logcpm_before_long $Samples <- as.factor(logcpm_before_long $Samples)

ggplot(logcpm_before_long , aes(x= Samples, y= counts, fill = Legend)) +
  geom_boxplot(width = 0.7) + theme_minimal() + 
  labs(title="Counts before normalization", x = '', y = "logCPM") +
  scale_fill_brewer(palette = 'PuOr')
```


Graph of logCPM after Normalization
```{r}
logcpm_after_long <- data.frame(logcpm_after)
#pivot the data frame into a long format
logcpm_after_long$GeneID <- rownames(logcpm_after_long)
names <- colnames(logcpm_after_long)[1: length(colnames(logcpm_after_long)) - 1]

logcpm_after_long  <- logcpm_after_long  %>% pivot_longer(cols= names ,
                    names_to= 'Samples',
                    values_to='counts')
logcpm_after_long$Legend <- ifelse(logcpm_after_long$Samples %in% brain, 'Brain', ifelse(logcpm_after_long$Samples %in% liver, 'Liver', 'Colon'))
logcpm_after_long $Samples <- as.factor(logcpm_after_long $Samples)

ggplot(logcpm_after_long , aes(x= Samples, y= counts, fill = Legend)) +
  geom_boxplot(width = 0.7) + theme_minimal() + 
  labs(title="Counts before normalization", x = '', y = "logCPM") +
  scale_fill_brewer(palette = 'PuOr')
```



## **Analysis for differially expressed genes**
We do not want to set an intercept since there is no base condition to which each sample can be compared to. 
```{r}
design <- model.matrix(~ 0 + group, data = y$samples)
colnames(design) <- levels(y$samples$group)
```

### MultiDimensional Scaling plot (MDS)
Using a MDS we can visualize our data in a 2 dimesional plot. The 2 dimension are the the Leading logFC dimension. And MDS plot uses the top 500 genes and compare their logFC to find similiarities between the samples. If the samples cluster toghether than we can assume they represent a similar condition. 

```{r}
plotMDS(logcpm, labels = group, main = 'Multidimensional scaling (MDS) plot', col=c(rep("orange",3),rep("mediumpurple",3),rep("black",3)))
```

Trying to see if sex can account for variability in the colon sample
```{r}

plotMDS(logcpm, labels = y$samples$sex, main = 'Multidimensional scaling (MDS) plot', col=c(rep("orange",3),rep("mediumpurple",3),rep("black",3)))
```


## **Biological Coefficient of Variation (BCV) plot**
This plot let us visualize the square root of dispersion (BCV) with respect to the average logCPM. In this plot we are able to see a a trend and is quite interesting that our samples doesn't show that much of a BCV given that the samples come form different tissues of different patients that can be different for age, sex, ecc…
```{r}
y <- estimateDisp(y, design)
plotBCV(y, main = 'Biological Coefficient of Variation (BCV) plot')
```

## **Differentially Expressed Genes** 
```{r}
fit <- glmQLFit(y, design)

# The order of the group is Brain, Colon, Liver since the order is alphabetical. 

# Colon vs Brain
qlfCB <- glmQLFTest(fit, contrast = c(-1, 1, 0))

# Liver vs Brain
qlfLB<- glmQLFTest(fit, contrast = c(-1, 0, 1))

# Liver vs Colon
qlfLC <- glmQLFTest(fit, contrast = c(0, -1, 1))
```

Extracting the full table of DE genes sorted by adjusted p-values
```{r}
results.CB <- topTags(qlfCB, n = 1000000, adjust.method = 'BH', sort.by = 'PValue', p.value = 1)
results.LB <- topTags(qlfLB, n = 1000000, adjust.method = 'BH', sort.by = 'PValue', p.value = 1)
results.LC <- topTags(qlfLC, n = 1000000, adjust.method = 'BH', sort.by = 'PValue', p.value = 1)
```


Saving the results in a excel file
```{r, results = 'hide'}
library(writexl)
write_xlsx(data.frame(results.CB), 'resultsCB.xlsx')
write_xlsx(data.frame(results.LB), 'resultsLB.xlsx')
write_xlsx(data.frame(results.LC), 'resultsLC.xlsx')
```

Quick summary for genes UP, DOWN and Not differential expressed for each comaprison: 
• Colon vs Brain
```{r}
#summary(decideTests(qlfCB, p.value = 0.05, lcf = 0))
summary(decideTests(qlfCB, p.value = 0.01, lcf = 1)) # different adjusted p.value and FC
```

• Liver vs Colon
```{r}
#summary(decideTests(qlfLB,  p.value = 0.05, lcf = 0))
summary(decideTests(qlfLB,  p.value = 0.01, lcf = 1)) # different adjusted p.value and FC
```

• Liver vs Brain
```{r}
#summary(decideTests(qlfLC,  p.value = 0.05, lcf = 0))
summary(decideTests(qlfLC,  p.value = 0.01, lcf = 1)) # different adjusted p.value and FC
```


## **Genes up regulated in one condition vs both conditions**
### Brain
```{r}
#Select the genes differentially expressed in one tissue against the other two
#Brain
sel_brain1 <- rownames(results.CB$table %>% filter(logFC < -1 & logCPM > 0 & FDR < 0.01))
sel_brain2 <- rownames(results.LB$table %>% filter(logFC < -1 & logCPM > 0 & FDR < 0.01))
topDEgenesBrain <- intersect(sel_brain1,sel_brain2)



table(startsWith(topDEgenesBrain, "RPL"))



maskBrain <- startsWith(topDEgenesBrain, "LOC") | startsWith(topDEgenesBrain,"MIR") | startsWith(topDEgenesBrain, "LINC") | startsWith(topDEgenesBrain, "SNORD")



topDEgenesBrain <- topDEgenesBrain[!maskBrain]
```
### Liver
```{r}
sel_liver1 <- rownames(results.LC$table %>% filter(logFC > 1 & logCPM > 0 & FDR < 0.01))
sel_liver2 <- rownames(results.LB$table %>% filter(logFC > 1 & logCPM > 0 & FDR < 0.01))
topDEgenesLiver <- intersect(sel_liver1,sel_liver2)



table(startsWith(topDEgenesLiver, "RPL"))



maskLiver <- startsWith(topDEgenesLiver, "LOC") | startsWith(topDEgenesLiver,"MIR") | startsWith(topDEgenesLiver, "LINC") | startsWith(topDEgenesLiver, "SNORD")



topDEgenesLiver <- topDEgenesLiver[!maskLiver]
```
### Colon
```{r}
sel_colon1 <- rownames(results.LC$table %>% filter(logFC < -1 & logCPM > 0 & FDR < 0.01))
sel_colon2 <- rownames(results.CB$table %>% filter(logFC > 1 & logCPM > 0 & FDR < 0.01))
topDEgenesColon <- intersect(sel_colon1,sel_colon2)



table(startsWith(topDEgenesColon, "RPL"))



maskColon <- startsWith(topDEgenesColon, "LOC") | startsWith(topDEgenesColon,"MIR") | startsWith(topDEgenesColon, "LINC") | startsWith(topDEgenesColon, "SNORD")



topDEgenesColon <- topDEgenesColon[!maskColon]

```

```{r}
write_xlsx(data.frame(topDEgenesBrain), 'topDEgenesBrain_full.xlsx')
write_xlsx(data.frame(topDEgenesLiver), 'topDEgenesLiver_full.xlsx')
write_xlsx(data.frame(topDEgenesColon), 'topDEgenesColon_full.xlsx')
```


## **“Functional” Enrichment Analysis **
We used ENRICHER to perform Functional enrichment analysis

Here you can find the page of enricher: 

[Brain_full](https://maayanlab.cloud/Enrichr/enrich?dataset=df549451f7fa6fdf82bb912e986297e8).
[Colon_full](https://maayanlab.cloud/Enrichr/enrich?dataset=10d7fba84a0ee22676eb62208ba00809).
[Liver_full](https://maayanlab.cloud/Enrichr/enrich?dataset=65967f81f0d99de87e70a7c7c2047e04).


## Wide analysis of a single gene
```{r}
assays(rse_brain)$TPM <- recount::getTPM(rse_brain)
assays(rse_liver)$TPM <- recount::getTPM(rse_liver)
assays(rse_colon)$TPM <- recount::getTPM(rse_colon)
```

```{r}
id_NAT8L  <- which(rowData(rse_brain)$gene_name ==  'NAT8L')
id_FOXF2 <-  which(rowData(rse_brain)$gene_name == 'FOXF2')
id_ALB  <- which(rowData(rse_brain)$gene_name ==  'ALB')
```

Now, find the most appropriate statistical test, and prove that the difference of expression for the gene is still significant if we consider all the samples, instead of just three per tissue.

Man Withney U test 
```{r, results= 'hide'}
library(dbplyr)
library(ggpubr)
library(rstatix)
```

```{r}
df_brain <- data.frame(TPM = assays(rse_brain)$TPM[id_NAT8L,],group="Brain") 
df_colon <- data.frame(TPM = assays(rse_colon)$TPM[id_NAT8L,],group="Colon") 
df_liver <- data.frame(TPM = assays(rse_liver)$TPM[id_NAT8L,],group="Liver") 
data_RNASE4 <- rbind(df_brain, df_colon, df_liver) 

res_kruskal <- data_RNASE4 %>% kruskal_test(TPM ~ group) 
res_kruskal 

pwc2 <- data_RNASE4 %>% wilcox_test(TPM ~ group, p.adjust.method = "BH") 
pwc2 

pwc <- pwc2 %>% add_xy_position(x = "group") #Auto-compute p-value label positions 
ggboxplot(data_RNASE4, x = "group", y = "TPM",outlier.shape = NA,width = 0.5,title="NAT8L expression across organs", fill = 'orange') + 
stat_pvalue_manual(pwc,y.position = c(700,700,700)) + #Add the p-values to the plot 
labs(subtitle = get_test_label(res_kruskal, detailed = TRUE),caption = get_pwc_label(pwc)) #test information (top:Kruskal, bottom:pairwise comparison)
```

```{r}
df_brain <- data.frame(TPM = assays(rse_brain)$TPM[id_FOXF2,],group="Brain") 
df_colon <- data.frame(TPM = assays(rse_colon)$TPM[id_FOXF2,],group="Colon") 
df_liver <- data.frame(TPM = assays(rse_liver)$TPM[id_FOXF2,],group="Liver") 
data_RNASE4 <- rbind(df_brain, df_colon, df_liver) 

res_kruskal <- data_RNASE4 %>% kruskal_test(TPM ~ group) 
res_kruskal 

pwc2 <- data_RNASE4 %>% wilcox_test(TPM ~ group, p.adjust.method = "BH") 
pwc2 

pwc <- pwc2 %>% add_xy_position(x = "group") #Auto-compute p-value label positions 
ggboxplot(data_RNASE4, x = "group", y = "TPM",outlier.shape = NA,width = 0.5,title="FOXF2 expression across organs", fill = 'seashell') + 
stat_pvalue_manual(pwc,y.position = c(300,300,300)) + #Add the p-values to the plot 
labs(subtitle = get_test_label(res_kruskal, detailed = TRUE),caption = get_pwc_label(pwc)) #test information (top:Kruskal, bottom:pairwise comparison)
```

```{r}
df_brain <- data.frame(TPM = assays(rse_brain)$TPM[id_ALB,],group="Brain") 
df_colon <- data.frame(TPM = assays(rse_colon)$TPM[id_ALB,],group="Colon") 
df_liver <- data.frame(TPM = assays(rse_liver)$TPM[id_ALB,],group="Liver") 
data_RNASE4 <- rbind(df_brain, df_colon, df_liver) 

res_kruskal <- data_RNASE4 %>% kruskal_test(TPM ~ group) 
res_kruskal 

pwc2 <- data_RNASE4 %>% wilcox_test(TPM ~ group, p.adjust.method = "BH") 
pwc2 

pwc <- pwc2 %>% add_xy_position(x = "group") #Auto-compute p-value label positions 
ggboxplot(data_RNASE4, x = "group", y = "TPM",outlier.shape = NA,width = 0.5,title="Albumin (ALB) expression across organs", fill = 'mediumpurple') + 
stat_pvalue_manual(pwc,y.position = c(130000, 130000, 130000)) + #Add the p-values to the plot 
labs(subtitle = get_test_label(res_kruskal, detailed = TRUE),caption = get_pwc_label(pwc)) #test information (top:Kruskal, bottom:pairwise comparison)
```
```{r}
id_ASGR1  <- which(rowData(rse_brain)$gene_name ==  'ASGR1')
df_brain <- data.frame(TPM = assays(rse_brain)$TPM[id_ASGR1,],group="Brain") 
df_colon <- data.frame(TPM = assays(rse_colon)$TPM[id_ASGR1,],group="Colon") 
df_liver <- data.frame(TPM = assays(rse_liver)$TPM[id_ASGR1,],group="Liver") 
data_RNASE4 <- rbind(df_brain, df_colon, df_liver) 

res_kruskal <- data_RNASE4 %>% kruskal_test(TPM ~ group) 
res_kruskal 

pwc2 <- data_RNASE4 %>% wilcox_test(TPM ~ group, p.adjust.method = "BH") 
pwc2 

pwc <- pwc2 %>% add_xy_position(x = "group") #Auto-compute p-value label positions 
ggboxplot(data_RNASE4, x = "group", y = "TPM",outlier.shape = NA,width = 0.5,title="ASGR1 expression across organs", fill = 'mediumpurple') + 
stat_pvalue_manual(pwc,y.position = c(1500, 1500, 1500)) + #Add the p-values to the plot 
labs(subtitle = get_test_label(res_kruskal, detailed = TRUE),caption = get_pwc_label(pwc)) #test information (top:Kruskal, bottom:pairwise comparison)
```




# **Clean Bulk Analysis**

### Loading the **data**
```{r, results= 'hide'}
rse_brain <- readRDS('rse_brain.RDS')
rse_liver <- readRDS('rse_liver.RDS')
rse_colon <- readRDS('rse_colon.RDS')
```

## Filtering out non canonical chromosomes
### In Brain
```{r}
canonical <- paste("chr", seq(1,22), sep="") 
canonical <- c(canonical, "chrX", "chrY") 

# Filter 
rse_brain<- rse_brain[ 
# Ribosomal RNA 
rowData(rse_brain)$gbkey != 'rRNA' & 
# Pseudogenes 
rowData(rse_brain)$gbkey != 'Gene' & 
# Exclude Non-canonical Chromosomes and Mitochondrial DNA 
rowRanges(rse_brain)@seqnames %in% canonical & 
# NAs 
!is.na(rowData(rse_brain)$gbkey), 
] 
```
### In Liver 
```{r}
# Filter 
rse_liver <- rse_liver[ 
# Ribosomal RNA 
rowData(rse_liver)$gbkey != 'rRNA' & 
# Pseudogenes 
rowData(rse_liver)$gbkey != 'Gene' & 
# Exclude Non-canonical Chromosomes and Mitochondrial DNA 
rowRanges(rse_liver)@seqnames %in% canonical & 
# NAs 
!is.na(rowData(rse_liver)$gbkey), 
] 
```
### And in Colon
```{r}
# Filter 
rse_colon <- rse_colon[ 
# Ribosomal RNA 
rowData(rse_colon)$gbkey != 'rRNA' & 
# Pseudogenes 
rowData(rse_colon)$gbkey != 'Gene' & 
# Exclude Non-canonical Chromosomes and Mitochondrial DNA 
rowRanges(rse_colon)@seqnames %in% canonical & 
# NAs 
!is.na(rowData(rse_colon)$gbkey), 
] 
```

Repeting the analysis: 
```{r}
assays(rse_brain)$counts <- transform_counts(rse_brain)
assays(rse_liver)$counts <- transform_counts(rse_liver)
assays(rse_colon)$counts <- transform_counts(rse_colon)
```

## **Filtering out rRNA, Pseudogenes and reads mapped to mitochondrial genes for Brain**
```{r}
# Drop reads mapped to rRNA
rse <- rse_brain[ 
rowData(rse_brain)$gbkey != 'rRNA' & 
!is.na(rowData(rse_brain)$gbkey),] 

# Drop reads mapped to Pseudogenes 
rse <- rse[ 
rowData(rse)$gbkey != 'Gene' & 
!is.na(rowData(rse)$gbkey),] 

# Drop reads mapped to mitochondrial genes 
rse_brain <- rse[rowRanges(rse)@seqnames != "chrM", ] 
```

Now for Liver 
```{r}
# Drop reads mapped to rRNA
rse <- rse_liver[ 
rowData(rse_liver)$gbkey != 'rRNA' & 
!is.na(rowData(rse_liver)$gbkey),] 

# Drop reads mapped to Pseudogenes 
rse <- rse[ 
rowData(rse)$gbkey != 'Gene' & 
!is.na(rowData(rse)$gbkey),] 

# Drop reads mapped to mitochondrial genes 
rse_liver <- rse[rowRanges(rse)@seqnames != "chrM", ] 
```

And finally for Colon
```{r}
# Drop reads mapped to rRNA
rse <- rse_colon[ 
rowData(rse_colon)$gbkey != 'rRNA' & 
!is.na(rowData(rse_colon)$gbkey),] 

# Drop reads mapped to Pseudogenes 
rse <- rse[ 
rowData(rse)$gbkey != 'Gene' & 
!is.na(rowData(rse)$gbkey),] 

# Drop reads mapped to mitochondrial genes 
rse_colon <- rse[rowRanges(rse)@seqnames != "chrM", ] 
```


## **Picking the replicates**. 
Selecting the replicates for each tissue:
```{r}
rse_brain_selected <- rse_brain[, c(90, 91, 92)]
rse_liver_selected <- rse_liver[, c(91, 94, 95)]
rse_colon_selected <- rse_colon[, c(90, 91, 92)]
```


## **Converting** coverage into counts 
```{r}
counts_brain_selected <- assays(rse_brain_selected)$counts
counts_liver_selected <- assays(rse_liver_selected)$counts
counts_colon_selected <- assays(rse_colon_selected)$counts
```

## **Count table** 
```{r}
count.table <- cbind(counts_brain_selected, counts_liver_selected, counts_colon_selected) 

colnames(count.table) <- c('Brain.90', 'Brain.91', 'Brain.92', 'Liver.91', 'Liver.94', 'Liver.95', 'Colon.90', 'Colon.91', 'Colon.92')

rownames(count.table) <- rowData(rse_brain_selected)$gene_name 

y <- DGEList(counts =  count.table)
```

## **Grouping** the replicates and storing **quality inforrmation**
```{r}
group <- as.factor(c('Brain', 'Brain', 'Brain', 'Liver', 'Liver', 'Liver', 'Colon', 'Colon', 'Colon'))

y$samples$group <- group
```

```{r}
# RIN 
y$samples$rin <- as.factor(c(
  colData(rse_brain_selected)$gtex.smrin, 
  colData(rse_liver_selected)$gtex.smrin, 
  colData(rse_colon_selected)$gtex.smrin))

# rRNA
y$samples$rRNA <- as.factor(c(
  colData(rse_brain_selected)$gtex.smrrnart, 
  colData(rse_liver_selected)$gtex.smrrnart, 
  colData(rse_colon_selected)$gtex.smrrnart))
# UNIQUELY MAPPED READS
y$samples$mapped <- as.factor(c(
  colData(rse_brain_selected)$'recount_qc.star.uniquely_mapped_reads_%_both',
  colData(rse_liver_selected)$'recount_qc.star.uniquely_mapped_reads_%_both', 
  colData(rse_colon_selected)$'recount_qc.star.uniquely_mapped_reads_%_both'))

# CHR % of reads mapping on the mitochondrial chromosome
y$samples$chrM <- as.factor(c(
  colData(rse_brain_selected)$'recount_qc.aligned_reads%.chrm',
  colData(rse_liver_selected)$'recount_qc.aligned_reads%.chrm', 
  colData(rse_colon_selected)$'recount_qc.aligned_reads%.chrm'))

# SLICE 
y$samples$slice <- as.factor(c(
  colData(rse_brain_selected)$gtex.smtsd, 
  colData(rse_liver_selected)$gtex.smtsd, 
  colData(rse_colon_selected)$gtex.smtsd))

# SEX 
y$samples$sex <- as.factor(c(
  colData(rse_brain_selected)$gtex.sex, 
  colData(rse_liver_selected)$gtex.sex, 
  colData(rse_colon_selected)$gtex.sex))

# AGE
y$samples$age <- as.factor(c(
  colData(rse_brain_selected)$gtex.age, 
  colData(rse_liver_selected)$gtex.age, 
  colData(rse_colon_selected)$gtex.age))
```



For a quick check we can look at how many genes have 0 counts
```{r}
table(rowSums(y$counts == 0) == 9)
```

## **Filtering by expression**
```{r}
keep.expressed <- filterByExpr(y, group = group)
y <- y [keep.expressed, , keep.lib.sizes = F]
```

## **Transforming counts** in Counts Per Milion in a Logarithmich scale (logCPM)
```{r}
logcpm_before <- cpm(y, log = T)
y <- calcNormFactors(y, method = 'TMM')
head(y)

logcpm <- cpm(y, log = T)
logcpm_after <- logcpm
```

We can visualize the normalization applied to the logCPM using box-plots

**Graph of logCPM before normalization**
```{r}
brain <- c('Brain.90', 'Brain.91', 'Brain.92')
liver <- c('Liver.91', 'Liver.94', 'Liver.95')
colon <- c('Colon.90', 'Colon.91', 'Colon.92')

logcpm_before_long <- data.frame(logcpm_before)
#pivot the data frame into a long format
logcpm_before_long$GeneID <- rownames(logcpm_before_long)
names <- colnames(logcpm_before_long)[1: length(colnames(logcpm_before_long)) - 1]

logcpm_before_long  <- logcpm_before_long  %>% pivot_longer(cols= names ,
                    names_to= 'Samples',
                    values_to='counts')
logcpm_before_long$Legend <- ifelse(logcpm_before_long$Samples %in% brain, 'Brain', ifelse(logcpm_before_long$Samples %in% liver, 'Liver', 'Colon'))
logcpm_before_long $Samples <- as.factor(logcpm_before_long $Samples)

ggplot(logcpm_before_long , aes(x= Samples, y= counts, fill = Legend)) +
  geom_boxplot(width = 0.7) + theme_minimal() + 
  labs(title="Counts before normalization", x = '', y = "logCPM") +
  scale_fill_brewer(palette = 'PuOr')
```


**Graph of logCPM after Normalization**
```{r}
logcpm_after_long <- data.frame(logcpm_after)
#pivot the data frame into a long format
logcpm_after_long$GeneID <- rownames(logcpm_after_long)
names <- colnames(logcpm_after_long)[1: length(colnames(logcpm_after_long)) - 1]

logcpm_after_long  <- logcpm_after_long  %>% pivot_longer(cols= names ,
                    names_to= 'Samples',
                    values_to='counts')
logcpm_after_long$Legend <- ifelse(logcpm_after_long$Samples %in% brain, 'Brain', ifelse(logcpm_after_long$Samples %in% liver, 'Liver', 'Colon'))
logcpm_after_long $Samples <- as.factor(logcpm_after_long $Samples)

ggplot(logcpm_after_long , aes(x= Samples, y= counts, fill = Legend)) +
  geom_boxplot(width = 0.7) + theme_minimal() + 
  labs(title="Counts before normalization", x = '', y = "logCPM") +
  scale_fill_brewer(palette = 'PuOr')
```



## **Analysis for differentially expressed genes**
```{r}
design <- model.matrix(~ 0 + group, data = y$samples)
colnames(design) <- levels(y$samples$group)
```

### MultiDimensional Scaling plot (MDS)
Using a MDS we can visualize our data in a 2 dimensional plot. The 2 dimension are the the Leading logFC dimension. And MDS plot uses the top 500 genes and compare their logFC to find similarities between the samples. If the samples cluster together than we can assume they represent a similar condition. 

```{r}
plotMDS(logcpm, labels = group, main = 'Multidimensional scaling (MDS) plot', col=c(rep("orange",3),rep("mediumpurple",3),rep("black",3)))
```

Trying to see if sex can account for variability in the colon sample
```{r}
plotMDS(logcpm, labels = y$samples$sex, main = 'Multidimensional scaling (MDS) plot', col=c(rep("orange",3),rep("mediumpurple",3),rep("black",3)))
```


### **Biological Coefficient of Variation (BCV) plot**
This plot let us visualize the square root of dispersion (BCV) with respect to the average logCPM. In this plot we are able to see a a trend and is quite interesting that our samples doesn't show that much of a BCV given that the samples come form different tissues of different patients that can be different for age, sex, ecc…
```{r}
y <- estimateDisp(y, design)
plotBCV(y, main = 'Biological Coefficient of Variation (BCV) plot')
```

### **Differentially Expressed Genes** 
```{r}
fit <- glmQLFit(y, design)

# The order of the group is Brain, Colon, Liver since the order is alphabetical. 

# Colon vs Brain
qlfCB <- glmQLFTest(fit, contrast = c(-1, 1, 0))

# Liver vs Brain
qlfLB<- glmQLFTest(fit, contrast = c(-1, 0, 1))

# Liver vs Colon
qlfLC <- glmQLFTest(fit, contrast = c(0, -1, 1))
```

Extracting the full table of DE genes sorted by adjusted p-values
```{r}
results.CB <- topTags(qlfCB, n = 1000000, adjust.method = 'BH', sort.by = 'PValue', p.value = 1)
results.LB <- topTags(qlfLB, n = 1000000, adjust.method = 'BH', sort.by = 'PValue', p.value = 1)
results.LC <- topTags(qlfLC, n = 1000000, adjust.method = 'BH', sort.by = 'PValue', p.value = 1)
```

```{r}
head(results.CB)
```

Saving the results in a excel file
```{r}
write_xlsx(data.frame(results.CB), 'resultsCB_cleaned.xlsx')
write_xlsx(data.frame(results.LB), 'resultsLB_cleaned.xlsx')
write_xlsx(data.frame(results.LC), 'resultsLC_cleaned.xlsx')
```

Quick summary for genes UP, DOWN and Not differential expressed for each comaprison: 
• Brain vs Colon
```{r}
#summary(decideTests(qlfCB, p.value = 0.05, lcf = 0))
summary(decideTests(qlfCB, p.value = 0.01, lcf = 1)) # different adjusted p.value and FC
```

• Liver vs Colon
```{r}
#summary(decideTests(qlfLB,  p.value = 0.05, lcf = 0))
summary(decideTests(qlfLB,  p.value = 0.01, lcf = 1)) # different adjusted p.value and FC
```

• Liver vs Brain
```{r}
#summary(decideTests(qlfLC,  p.value = 0.05, lcf = 0))
summary(decideTests(qlfLC,  p.value = 0.01, lcf = 1)) # different adjusted p.value and FC
```


# genes up regulated in one condition vs both conditions
```{r}

#Select the genes differentially expressed in one tissue against the other two
#Brain
sel_brain1 <- rownames(results.CB$table %>% filter(logFC < -1 & logCPM > 0 & FDR < 0.01))
sel_brain2 <- rownames(results.LB$table %>% filter(logFC < -1 & logCPM > 0 & FDR < 0.01))
topDEgenesBrain <- intersect(sel_brain1,sel_brain2)



table(startsWith(topDEgenesBrain, "RPL"))



maskBrain <- startsWith(topDEgenesBrain, "LOC") | startsWith(topDEgenesBrain,"MIR") | startsWith(topDEgenesBrain, "LINC") | startsWith(topDEgenesBrain, "SNORD")



topDEgenesBrain <- topDEgenesBrain[!maskBrain]
```
#Liver
```{r}
sel_liver1 <- rownames(results.LC$table %>% filter(logFC > 1 & logCPM > 0 & FDR < 0.01))
sel_liver2 <- rownames(results.LB$table %>% filter(logFC > 1 & logCPM > 0 & FDR < 0.01))
topDEgenesLiver <- intersect(sel_liver1,sel_liver2)



table(startsWith(topDEgenesLiver, "RPL"))



maskLiver <- startsWith(topDEgenesLiver, "LOC") | startsWith(topDEgenesLiver,"MIR") | startsWith(topDEgenesLiver, "LINC") | startsWith(topDEgenesLiver, "SNORD")



topDEgenesLiver <- topDEgenesLiver[!maskLiver]
```
#Colon
```{r}
sel_colon1 <- rownames(results.LC$table %>% filter(logFC < -1 & logCPM > 0 & FDR < 0.01))
sel_colon2 <- rownames(results.CB$table %>% filter(logFC > 1 & logCPM > 0 & FDR < 0.01))
topDEgenesColon <- intersect(sel_colon1,sel_colon2)



table(startsWith(topDEgenesColon, "RPL"))



maskColon <- startsWith(topDEgenesColon, "LOC") | startsWith(topDEgenesColon,"MIR") | startsWith(topDEgenesColon, "LINC") | startsWith(topDEgenesColon, "SNORD")



topDEgenesColon <- topDEgenesColon[!maskColon]
```

```{r}
write_xlsx(data.frame(topDEgenesBrain), 'topDEgenesBrain_cleaned.xlsx')
write_xlsx(data.frame(topDEgenesLiver), 'topDEgenesLiver_cleaned.xlsx')
write_xlsx(data.frame(topDEgenesColon), 'topDEgenesColon_cleaned.xlsx')
```

## **“Functional” Enrichment Analysis **
We used ENRICHER to perform Functional enrichment analysis

Here you can find the page of enricher: 

[Brain_clean](https://maayanlab.cloud/Enrichr/enrich?dataset=252bbcc30400a9f30df72911cc52f301).
[Colon_clean](https://maayanlab.cloud/Enrichr/enrich?dataset=91bf09d66ddc7cca16487c5273b4ef5f).
[Liver_clean](https://maayanlab.cloud/Enrichr/enrich?dataset=9a0b5a68356caf4c7802395a96cea2bc).


### WIDE ANALYSIS OF A SINGLE GENE
```{r}
assays(rse_brain)$TPM <- recount::getTPM(rse_brain)
assays(rse_liver)$TPM <- recount::getTPM(rse_liver)
assays(rse_colon)$TPM <- recount::getTPM(rse_colon)
```

```{r}
id_PDE6B  <- which(rowData(rse_brain)$gene_name == 'PDE6B')
id_TEAD3 <-  which(rowData(rse_brain)$gene_name == 'TEAD3')
id_ASGR1  <- which(rowData(rse_brain)$gene_name ==  'ASGR1')
```

Now, find the most appropriate statistical test, and prove that the difference of expression for the gene is still significant if we consider all the samples, instead of just three per tissue.

Man Withney U test 
```{r}
df_brain <- data.frame(TPM = assays(rse_brain)$TPM[id_PDE6B,],group="Brain") 
df_colon <- data.frame(TPM = assays(rse_colon)$TPM[id_PDE6B,],group="Colon") 
df_liver <- data.frame(TPM = assays(rse_liver)$TPM[id_PDE6B,],group="Liver") 
data_RNASE4 <- rbind(df_brain, df_colon, df_liver) 

library(rstatix)
res_kruskal <- data_RNASE4 %>% kruskal_test(TPM ~ group) 
res_kruskal 

pwc2 <- data_RNASE4 %>% wilcox_test(TPM ~ group, p.adjust.method = "BH") 
pwc2 

pwc <- pwc2 %>% add_xy_position(x = "group") #Auto-compute p-value label positions 
ggboxplot(data_RNASE4, x = "group", y = "TPM",outlier.shape = NA,width = 0.5,title="PDE6B expression across organs", fill = 'orange', ylim = c(0, 60)) + 
stat_pvalue_manual(pwc,y.position = c(60, 60, 60)) + #Add the p-values to the plot 
labs(subtitle = get_test_label(res_kruskal, detailed = TRUE),caption = get_pwc_label(pwc)) #test information (top:Kruskal, bottom:pairwise comparison)
```

```{r}
df_brain <- data.frame(TPM = assays(rse_brain)$TPM[id_TEAD3,],group="Brain") 
df_colon <- data.frame(TPM = assays(rse_colon)$TPM[id_TEAD3,],group="Colon") 
df_liver <- data.frame(TPM = assays(rse_liver)$TPM[id_TEAD3,],group="Liver") 
data_RNASE4 <- rbind(df_brain, df_colon, df_liver) 

res_kruskal <- data_RNASE4 %>% kruskal_test(TPM ~ group) 
res_kruskal 

pwc2 <- data_RNASE4 %>% wilcox_test(TPM ~ group, p.adjust.method = "BH") 
pwc2 

pwc <- pwc2 %>% add_xy_position(x = "group") #Auto-compute p-value label positions 
ggboxplot(data_RNASE4, x = "group", y = "TPM",outlier.shape = NA,width = 0.5,title="TEAD3 expression across organs", fill = 'seashell', ylim = c(0, 700)) + 
stat_pvalue_manual(pwc,y.position = c(700,700,700)) + #Add the p-values to the plot 
labs(subtitle = get_test_label(res_kruskal, detailed = TRUE),caption = get_pwc_label(pwc)) #test information (top:Kruskal, bottom:pairwise comparison)
```

```{r}
df_brain <- data.frame(TPM = assays(rse_brain)$TPM[id_ASGR1,],group="Brain") 
df_colon <- data.frame(TPM = assays(rse_colon)$TPM[id_ASGR1,],group="Colon") 
df_liver <- data.frame(TPM = assays(rse_liver)$TPM[id_ASGR1,],group="Liver") 
data_RNASE4 <- rbind(df_brain, df_colon, df_liver) 

res_kruskal <- data_RNASE4 %>% kruskal_test(TPM ~ group) 
res_kruskal 

pwc2 <- data_RNASE4 %>% wilcox_test(TPM ~ group, p.adjust.method = "BH") 
pwc2 

pwc <- pwc2 %>% add_xy_position(x = "group") #Auto-compute p-value label positions 
ggboxplot(data_RNASE4, x = "group", y = "TPM",outlier.shape = NA,width = 0.5,title="ASGR1 expression across organs", fill = 'mediumpurple') + 
stat_pvalue_manual(pwc,y.position = c(2000, 2000, 2000)) + #Add the p-values to the plot 
labs(subtitle = get_test_label(res_kruskal, detailed = TRUE),caption = get_pwc_label(pwc)) #test information (top:Kruskal, bottom:pairwise comparison)
```


FORUM 
```{r}
# I was curious to infer the sex of the donor given the reads of the tissue sample.
# Since nobody posted about it, I was curious to provide my solution. 
# Please feel free to comment and ask me if something seems unclear. 
# (also please double check my code in case I made some mistake)

# Using a small data frame (explorative.df) to make some explorative analysis in Liver:
# We need to store both the sex of the donor given by Genotype-Tissue Expression (GTEx) project and the percentage of aligned reads for chromosome Y (given by Recount3). 
explorative.df <- data.frame(gtex.sex = rse_liver$gtex.sex , perc.chrmY = rse_liver$`recount_qc.aligned_reads%.chry`)

# Assuming that if a donor is female, it will have no reads aligned to the chromosome Y we can try to infer the sex of all the donors using a binary choice. 

number_of_patients = length(explorative.df$gtex.sex)

for (patient_i in seq(1 : number_of_patients) ){ 
    explorative.df[patient_i, 'inferred']  <- ifelse(explorative.df[patient_i, 'perc.chrmY'] == 0, 'Female', 'Male')
}

# This way we can store the inferred sex and check if there are some problems. 
# If our assumption is correct, than we would have labeled all the Male and Female correctly, without any ambiguity.
False_Males <- nrow(explorative.df[which(explorative.df$inferred == 'Male' & explorative.df$gtex.sex == 2),]) 
True_Males <- nrow(explorative.df[which(explorative.df$inferred == 'Male' & explorative.df$gtex.sex == 1),]) 

#But both sex 1 (176 in total) and  2 (31 in total) are labeled as Male.

#This is mainly due to the fact that X and Y chromosomes share three high sequence homology regions. 
#"100% sequence identity for the pseudoautosomal regions PARs, PAR1, and PAR2, and ~ 99% sequence homology in the X-transposed region XTR" citation.1 

# This may lead to a non 0 percent of aligned reads on the Y chromosomes in female donors. 

# In the end I will suggest to define a reasonable threshold around 0 to label your donor as Female or Male. Also is important to note that the threshold may vary greatly depending on the tissue (as an example in brain is much easier to define the threshold since almost all the 2 have 0 percentage of mapping reads on the Y chromosome, identifying once for all Females as 2 in the GTEx project.)



# If you are curious this articles, citation.1, goes in depth in how the sex of the donors can alter results in differential gene expression analysis and explain much better what are possible problems of using the percentage of aligned reads on the chromosome Y to infer the sex of the donor.

# citation.1
# Reference genome and transcriptome informed by the sex chromosome complement of the sample increase ability to detect sex differences in gene expression from RNA-Seq data.  https://doi.org/10.1186/s13293-020-00312-9 
```
```{r}
# We can also use a boxplot to visualize the distribution of the  percentage of mapped reads on the Y chromsomes in Females donors

library(ggpubr)
ggboxplot(explorative.df[explorative.df$gtex.sex == 2, ], x = 'gtex.sex' , y = 'perc.chrmY' , ylab = '% of alinged reads', xlab = 'Female donors',  width = 0.1, fill = 'purple', title = 'Distribution of % alinged reads in Liver', ylim = c(0, 0.025)) 
```



